gk/main.c

```diff
@@ -16,22 +16,310 @@
+ * (1) compute the priority of the packet.
+ * (2) encapsulate the packet as a request.
+ * (3) put this encapsulated packet in the request queue.
+ */
+static int
+gk_process_request(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+	uint8_t priority = priority_from_delta_time(now,
+			fe->u.request.last_packet_seen_at);
+
+	fe->u.request.last_packet_seen_at = now;
+	if (priority <= fe->u.request.last_priority &&
+			fe->u.request.allowance > 0) {
+		fe->u.request.allowance--;
+		priority = fe->u.request.last_priority;
+	} else {
+		fe->u.request.last_priority = priority;
+		fe->u.request.allowance = START_ALLOWANCE - 1;
+	}
+
+	if (fe->u.request.allowance <= 0) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"gk: the allowance for the request is 0, need to be dropped!\n");
+		return drop_packet(packet->pkt);
+	}
+
+	/*
+	 * Adjust @priority for the DSCP field.
+	 * DSCP 0 for legacy packets; 1 for granted packets; 
+	 * 2 for capability renew; 3-63 for requests.
+	 */
+	priority += 3;
+	if (unlikely(priority > 63))
+		priority = 63;
+
+	/* The assigned priority is @priority. */
+	
+	/* TODO Encapsulate the packet as a request. */
+
+	/* TODO Put this encapsulated packet in the request queue. */
+
+	return 0;
+}
+
+static uint64_t
+time_from_seconds(uint64_t time)
+{
+	return (cycles_per_sec * time);
+}
+
+static int
+gk_process_granted(struct flow_entry *fe, struct ipacket *packet)
+{
+	bool renew_cap;
+	uint64_t now = rte_rdtsc();
+	struct rte_mbuf *pkt = packet->pkt;
+
+	/* 
+	 * When a flow entry is at granted state, the GK block processes 
+	 * that entry as follows:
+	 */
+	if (now >= fe->u.granted.cap_expire_at) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	if (now >= fe->u.granted.budget_renew_at) {
+		fe->u.granted.budget_renew_at = now + time_from_seconds(1);
+		fe->u.granted.budget_byte = fe->u.granted.tx_rate_kb_cycle * 1024;
+	}
+
+	if (pkt->data_len > fe->u.granted.budget_byte)
+		return drop_packet(pkt);
+
+	fe->u.granted.budget_byte -= pkt->data_len;
+	renew_cap = now >= fe->u.granted.send_next_renewal_at;
+	if (renew_cap)
+		fe->u.granted.send_next_renewal_at = now +
+			fe->u.granted.renewal_step_cycle;
+
+	/*
+	 * TODO Encapsulate packet as a granted packet,
+	 * mark it as a capability renewal request if @renew_cap is true,
+	 * enter destination according to @fe->u.granted.grantor_id,
+	 * and put the encapsulated packet in the granted queue.
+	 */
+
+	return 0;
+}
+
+static int
+gk_process_declined(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+
+	/*
+	 * When a flow entry is at declined state, the GK block processes
+	 * that entry as follows:
+	 */
+	if (unlikely(now >= fe->u.declined.expire_at)) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	return drop_packet(packet->pkt);
+}
+
+static int
+setup_gk_instance(unsigned int block_idx, struct gk_config *gk_conf)
```

Comment:
I think ``uint32_t`` and ``unsigned int`` are equivalent, so I used them interchangeably. I will keep them consistent.